Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/J2ObjCParser.kt

```diff
@@ -0,0 +1,218 @@
+                           signature: String?,
+                           exceptions: Array<out String>?): MethodVisitor? {
+
+    val methodBuilder = MethodBuilder(parameterNames)
+    methodDescriptors.add(Triple(name, descriptor, access))
+    return methodBuilder
+  }
+
+  /**
+   * Generates an ObjCClass out of data collected while visiting
+   *
+   * @return An ObjCClass that matches a Java class
+   */
+  fun buildClass(): ObjCClass {
+    val methods = (methodDescriptors zip parameterNames).map { buildClassMethod(it.first.first, it.first.second, it.first.third, it.second)}
+
+    val generatedClass = ObjCClassImpl(
+      name = className,
+      isForwardDeclaration = false,
+      binaryName = null,
+      location = Location(HeaderId("")) // Leaving headerId empty for now
+    )
+    generatedClass.methods.addAll(methods)
+    generatedClass.baseClass = ObjCClassImpl(
+      name = "NSObject",
+      binaryName = null,
+      isForwardDeclaration = false,
+      location = Location(headerId = HeaderId("usr/include/objc/NSObject.h")) // TODO: When implementing inheritance check for proper base class
+    )
+    return generatedClass
+  }
+
+  /**
+   * Creates a ObjCMethod out of method data and parameter names
+   *
+   * @param methodName Name of method
+   * @param methodDesc Descriptor of method
+   * @param access Integer that describes access level of this method
+   * @param paramNames List of parameter names of this method taken from MethodBuilder
+   */
+  private fun buildClassMethod(methodName: String, methodDesc: String, access: Int, paramNames: List<String>): ObjCMethod {
+    if (methodName == "<init>") {
+      return ObjCMethod(
+        selector = "init",
+        encoding = "[]",
+        parameters = listOf<Parameter>(), // TODO: Support constructor arguments
+        returnType = ObjCInstanceType(nullability = ObjCPointer.Nullability.Unspecified),
+        isVariadic = false,
+        isClass = false,
+        nsConsumesSelf = true,
+        nsReturnsRetained = true,
+        isOptional = false,
+        isInit = true,
+        isExplicitlyDesignatedInitializer = false)
+    } else {
+      val selector = StringBuilder(methodName)
+      val methodParameters = parseMethodParameters(methodDesc, paramNames)
+      val methodReturnType = parseMethodReturnType(methodDesc)
+      if (methodParameters.size >= 1) methodParameters.subList(1,methodParameters.size).forEach { selector.append(":" + it.name) }
+      return ObjCMethod(
+        selector = if (methodParameters.size > 1) "$selector:" else if (methodParameters.size == 1) "$methodName:" else methodName,
+        encoding = "[]", //TODO: Implement encoding properly
+        parameters = methodParameters,
+        returnType = methodReturnType,
+        isVariadic = false,
+        isClass = access == Opcodes.ACC_PUBLIC + Opcodes.ACC_STATIC, // TODO: Currently only handles Public instance and Public static methods, true when static, false when instance
```

Comment:
technically you'll want to use "ACC_PUBLIC or ACC_STATIC" instead of plus, since this is a bitwise operation

Reply:
Done.